File: data_structures/Python/Binary_Tree.py
from Queue import Queue

class Binary_Tree:
    ##Inner class for each node, all values defined in add method
    class Node:
        def __init__(self,dat):
            self.data = dat
            self.left = None
            self.right = None
            self.bin_id = '' ##0 is a branch left, 1 is a branch right
    
    def __init__(self):
        self.root = None

    def add_node(self,data):
        
        new_node = Binary_Tree.Node(data)
        current_node = self.root

        ##Binary tree is empty
        if current_node == None:
            self.root = new_node

        else:
            while current_node != None:
                
                previous = current_node
                if current_node.data > new_node.data:
                    current_node = current_node.left
                    new_node.bin_id += '0'

                elif current_node.data < new_node.data:
                    current_node = current_node.right
                    new_node.bin_id += '1'

            if new_node.data < previous.data:
                previous.left = new_node
            elif new_node.data > previous.data:
                previous.right = new_node
                

    def delete_node(self,data):
        current_node = self.root
        while current_node != None and current_node.data != data:
            previous = current_node
            if data < current_node.data:
                current_node = current_node.left
            
            elif data > current_node.data:
                current_node = current_node.right
        ##Node to delete does not exist
        if current_node == None:
            return False
        
        ##Node exists and has no children
        if current_node.left == None and current_node.right == None:
            ##Case for deleting root
            if current_node == self.root:
                self.root = None
            elif previous.data > current_node.data:
                previous.left = None
            else:
                previous.right = None
        
        ##Node exists and has one child - left
        elif current_node.right == None:
            ##Case for deleting root
            if current_node == self.root:
                self.root = current_node.left
            elif previous.data > current_node.data:
                previous.left = current_node.left
            else:
                previous.right = current_node.left
        
        ##Node exists and has one child - Right
        elif current_node.left == None:
            ##Case for deleting root
            if current_node == self.root:
                self.root = current_node.right
            elif previous.data > current_node.data:
                previous.left = current_node.right
            else:
                previous.right = current_node.right

        ##Node exists and has two children
        else:
            right_node = current_node.right
            ##Right node has  a left subtree
            if right_node.left != None:
                ##Find smallest leaf node in right_node.left subtree
                smallest = right_node
                while smallest.left != None:
                    previous = smallest
                    smallest = smallest.left
                current_node.data = smallest.data
                previous.left = None
            ##Right node has no left subtree - it is the least value
            else:
                current_node.data = right_node.data
                current_node.right = right_node.right
        return True

    ##Call with root as a paramater
    def binary_search(self,current_node,value):
        if current_node == None or current_node.data == value:
            return current_node
        elif value < current_node.data:
            return self.binary_search(current_node.left,value)
        elif value > current_node.data:
            return self.binary_search(current_node.right,value)
        
    ##Call with root as parameter - modifies data_output in place
    ##Node Left Right
    def pre_order(self,current_node, data_output):
        if current_node != None:
            data_output.append([current_node.data,current_node.bin_id])
        
        if current_node.left != None:
            self.pre_order(current_node.left, data_output)
        if current_node.right != None:
            self.pre_order(current_node.right, data_output)

    ##Call with root as parameter - modifies data_output in place
    ##Left Node Right
    def in_order(self,current_node, data_output):
        if current_node != None:

            if current_node.left != None:
                self.in_order(current_node.left, data_output)

            data_output.append([current_node.data,current_node.bin_id])

            if current_node.right != None:
                self.in_order(current_node.right, data_output)

    ##Call with root as parameter - modifies data_output in place
    ##Left Right Node
    def post_order(self,current_node, data_output):
        if current_node != None:

            if current_node.left != None:
                self.post_order(current_node.left, data_output)
            if current_node.right != None:
                self.post_order(current_node.right, data_output)
            
            data_output.append([current_node.data,current_node.bin_id])
    

    def breadth_first(self):
        current_node = self.root
        q = Queue()
        data_output = []
        while current_node != None:
            ##data_output.append([current_node.data,current_node.bin_id])
            data_output.append(current_node.data)
            if current_node.left != None:
                q.enqueue(current_node.left)
            if current_node.right != None:
                q.enqueue(current_node.right)
            
            
            current_node = q.dequeue()
        return data_output

    ##Call with current_node as the root
    def reverse(self,current_node):
        if current_node == None:
            return None
        current_node.left, current_node.right = current_node.right, current_node.left
        new_id = ''
        for b in current_node.bin_id:
            new_id += '0' if b == '1' else '1'
        current_node.bin_id = new_id
        self.reverse(current_node.left)
        self.reverse(current_node.right)
    
    ##Call with self.root as node
    def get_height(self,node):
        if node != None:
            return 1 + max(self.get_height(node.left),self.get_height(node.right))
        else:
            return -1

    ##Get to the top-down view of the binary tree - i.e. the top node at each horizontal width
    def get_top_view(self):
        n_i_v = {}
        hd = 0
        height = 0
        n_i_v[0] = [height,self.root]
        height+=1
        ##To the left
        self.buildView(self.root.left,n_i_v,hd-1,height)
        ##To the right
        self.buildView(self.root.right,n_i_v,hd+1,height)
        nodes = [value[1].data for (key,value) in sorted(n_i_v.items())]
        return sorted(nodes)

    ##Recursive function for get_top_view()
    def buildView(self,node,n_i_v,hd,height):
        if node == None:
            return 
        try:
            test = n_i_v[hd]
            if test[0] > height:
                n_i_v[hd] = [height,node]
        except:
            n_i_v[hd] = [height,node]
        height+=1
        ##To the left
        self.buildView(node.left,n_i_v,hd-1,height)
        ##To the right
        self.buildView(node.right,n_i_v,hd+1,height)




def show():
    tree = Binary_Tree()
    vals = [25, 39, 46, 20, 3, 23, 35, 32, 8, 16, 12, 33, 36, 31, 49, 41, 2, 4, 6, 9, 17, 34, 24, 13, 0, 42, 37, 40, 44, 30]
    for val in vals:
        tree.add_node(val)
    '''
    ##print("\nRoot: ", tree.root.data)

    pre_out = []
    ##tree.pre_order(tree.root, pre_out)
    ##print("\nPre-Order:", [i[0] for i in pre_out])

    in_out = []
    tree.in_order(tree.root, in_out)
    print("\nIn-Order:",[i[0] for i in in_out])
    tree.delete_node(23)
    in_out = []
    tree.in_order(tree.root, in_out)
    print("\nIn-Order:",[i[0] for i in in_out])


    post_out = []
    tree.post_order(tree.root, post_out)
    print("\nPost-Order:",[i[0] for i in post_out])
    '''
    breadth_out = tree.breadth_first()
    print("\nBreadth-First:",breadth_out)
    '''
    height = tree.get_height(tree.root)
    print(height)

    six_node = tree.binary_search(tree.root,6)
    print("\nBinary Search:",six_node.data,six_node.bin_id)

    tree.reverse(tree.root)
    print("\nReversed Breadth-First:",tree.breadth_first())
    top_view = tree.get_top_view()
    print(top_view)
    '''
show()

File: data_structures/Java/BinaryTree.java
package Java;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Vector;
import Queue;

public class BinaryTree {

    private Node head;
    private int size;

    BinaryTree() {
        head = null;
    }
    
    static class Node {
        private int value;
        private Node left;
        private Node right;

        Node(int value) {
            this.value = value;
            left = null;
            right = null;
        }

        public int getValue() {
            return value;
        }

        public Node getLeft() {
            return left;
        }

        public Node getRight() {
            return right;
        }

        public void setLeft(Node node) {
            left = node;
        }

        public void setRight(Node node) {
            right = node;
        }

    }

    public static void main(String[] args) {
        int[] data = {25, 39, 46, 20, 3, 23, 35, 32, 8, 16, 12, 33, 36, 31, 49, 41, 2, 4, 6, 9, 17, 34, 24, 13, 0, 42, 37, 40, 44, 30};
        BinaryTree bin_tree = new BinaryTree();

        for (int i : data){
            bin_tree.addNode(i);
        }
        Integer[] in_order = bin_tree.getInOrder();
        Integer[] pre_order = bin_tree.getPreOrder();
        Integer[] post_order = bin_tree.getPostOrder();
        System.out.println("In order: " + Arrays.toString(in_order));
        System.out.println("Pre order: " + Arrays.toString(pre_order));
        System.out.println("Post order: " + Arrays.toString(post_order));

        Integer[] bfs = bin_tree.breadth_first();
        System.out.println("Breadth first: " + Arrays.toString(bfs));
        // boolean success = bin_tree.deleteNode(23);
        // if (success == true) {
        //     in_order = bin_tree.getInOrder();
        //     System.out.println(in_order);
        // }
        // bin_tree.reverse();
        // int[] rev_in_order = bin_tree.getInOrder();
        // System.out.println(rev_in_order);
        // int height = bin_tree.getHeight();
        // System.out.println(height);
    }

    public void addNode(int new_value) {
        size = size+1;
        Node new_node = new Node(new_value);
        Node current_node = head;
        Node previous_node = current_node;
        //Binary Tree is current empty
        if (current_node == null) {
            head = new_node;
        } else {
            while (current_node != null) {
                previous_node = current_node;
                current_node = (new_value > current_node.value) ? current_node.right : current_node.left;
            }
            if (new_value < previous_node.value) {
                previous_node.setLeft(new_node);
            } else if (new_value > previous_node.value) {
                previous_node.setRight(new_node);
            }
        }
    }

    private void genInOrder(Node current_node, ArrayList<Integer> values) {
        if (current_node != null) {
            if (current_node.left != null) {
                genInOrder(current_node.left, values);
            }
            values.add(current_node.value);
            if (current_node.right != null) {
                genInOrder(current_node.right, values);
            }
        }
    }

    public Integer[] getInOrder() {
        ArrayList<Integer> output = new ArrayList<Integer>();
        //Vector<Integer> output = new Vector<>();
        genInOrder(head, output);
        Integer[] out = output.toArray(new Integer[0]);
        return out;
    }

    //Pre-Order generating [root, pre-order] values
    private void genPreOrder(Node current_node, ArrayList<Integer> values) {
        if (current_node != null) {
            values.add(current_node.value);
            if (current_node.left != null) {
                genPreOrder(current_node.left, values);
            }
            if (current_node.right != null) {
                genPreOrder(current_node.right, values);
            }
        }
    }

    public Integer[] getPreOrder() {
        ArrayList<Integer> output = new ArrayList<Integer>();
        genPreOrder(head, output);
        Integer[] out = output.toArray(new Integer[0]);
        return out;
    }

    private void genPostOrder(Node current_node, ArrayList<Integer> values) {
        if (current_node != null) {
            if (current_node.left != null) {
                genPostOrder(current_node.left, values);
            }
            if (current_node.left != null) {
                genPostOrder(current_node.right, values);
            }
            values.add(current_node.value);
        }
    }

    public Integer[] getPostOrder() {
        ArrayList<Integer> output = new ArrayList<Integer>();
        genPostOrder(head,output);
        Integer[] out = output.toArray(new Integer[0]);
        return out;
    }

    //add bfs when queue can be imported
    public Integer[] breadth_first() {
        Node current = head;
        Queue<Node> q = new Queue<Node>();
        ArrayList<Integer> output = new ArrayList<Integer>();
        while (current != null) {
            output.add(current.value);

            if (current.left != null) {
                q.enqueue(current.left);
            }
            if (current.right != null) {
                q.enqueue(current.right);
            }

            current = q.dequeue();
        }
        Integer[] out = output.toArray(new Integer[0]);
        return out;
    }

    private void reverseTree(Node current_node) {
        if (current_node == null) {
            return;
        }
        Node temp_left = current_node.left;
        current_node.setLeft(current_node.right);
        current_node.setRight(temp_left);
        reverseTree(current_node.left);
        reverseTree(current_node.right);
    }

    public void reverse() {
        reverseTree(head);
    }

    private int genHeight(Node current_node) {
        if (current_node != null) {
            return 1 + Math.max(genHeight(current_node.left),genHeight(current_node.right));   
        } else {
            return -1;
        }
    }

    public int getHeight() {
        int height = genHeight(head);
        return height;
    }

    public boolean deleteNode(int val) {
        boolean success;
        Node current = head;
        Node previous = current;
        while ((current != null) && (current.value != val)) {
            previous = current;
            if (val < current.value) {
                current = current.left;
            } else if (val > current.value) {
                current = current.right;
            }
        }
        if (current == null) {
            //Node to delete does not exist
            success = false;

        } else if ((current.left == null) && (current.right) == null) {
            //Node has no children
            if (current == head) {
                //Case for deleting head of tree
                head = null;
            } else if (previous.value > current.value) {
                previous.left = null;                
            } else {
                previous.right = null;
            }
            success =  true;

        } else if (current.right == null) {
            //Node only has a left child
            if (current == head) {
                //Case for deleting head of tree
                head = current.left;
            } else if (previous.value > current.value) {
                previous.left = current.left;
            } else {
                previous.right = current.right;
            }
            success =  true;

        } else if (current.left == null) {
            //Node only has a right child
            if (current == head) {
                //Case for deleting head of tree
                head = current.right;
            } else if (previous.value > current.value){
                previous.left = current.right;
            } else {
                previous.right = current.right;
            }
            success = true;

        } else {
            //Node exists and has two children - Hibbard deletion
            Node rightNode = current.right;
            if (rightNode.left != null) {
                //Right node has a left subtree
                Node min = rightNode;
                while (min.left != null) {
                    //Find smallest node in subtree
                    previous = min;
                    min = min.left;
                }
                current.value = min.value; //Overwrite deleting node
                previous.left = null; //Erase moved node
            } else {
                //Right node has no subtree
                current.value = rightNode.value;
                current.right = rightNode.right;
            }
            success = true;
        }
        return success;
    }
}


